streams/src/test/java/org/apache/kafka/streams/integration/RegexSourceIntegrationTest.java

```diff
@@ -0,0 +1,353 @@
+
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.junit.Assert.assertThat;
+
+/**
+ * End-to-end integration test based on using regex and named topics for creating sources, using
+ * an embedded Kafka cluster.
+ */
+
+public class RegexSourceIntegrationTest {
+    @ClassRule
+    public static final EmbeddedSingleNodeKafkaCluster CLUSTER = new EmbeddedSingleNodeKafkaCluster();
+
+    private static final String TOPIC_1 = "topic-1";
+    private static final String TOPIC_2 = "topic-2";
+    private static final String TOPIC_3 = "topic-3";
+    private static final String TOPIC_A = "topic-A";
+    private static final String TOPIC_C = "topic-C";
+    private static final String TOPIC_Y = "topic-Y";
+    private static final String TOPIC_Z = "topic-Z";
+    private static final String FA_TOPIC = "fa";
+    private static final String FOO_TOPIC = "foo";
+
+    private static final int FIRST_UPDATE = 0;
+    private static final int SECOND_UPDATE = 1;
+
+    private static final String DEFAULT_OUTPUT_TOPIC = "outputTopic";
+
+    @BeforeClass
+    public static void startKafkaCluster() throws Exception {
+        CLUSTER.createTopic(TOPIC_1);
+        CLUSTER.createTopic(TOPIC_2);
+        CLUSTER.createTopic(TOPIC_A);
+        CLUSTER.createTopic(TOPIC_C);
+        CLUSTER.createTopic(TOPIC_Y);
+        CLUSTER.createTopic(TOPIC_Z);
+        CLUSTER.createTopic(FA_TOPIC);
+        CLUSTER.createTopic(FOO_TOPIC);
+
+    }
+
+    @Test
+    public void testRegexMatchesTopicsAWhenCreated() throws Exception {
+
+        final Serde<String> stringSerde = Serdes.String();
+
+        Properties streamsConfiguration = getStreamsConfig();
+        StreamsConfig streamsConfig = new StreamsConfig(streamsConfiguration);
+
+        CLUSTER.createTopic("TEST-TOPIC-1");
+
+        KStreamBuilder builder = new KStreamBuilder();
+
+        KStream<String, String> pattern1Stream = builder.stream(Pattern.compile("TEST-TOPIC-\\d"));
+
+        pattern1Stream.to(stringSerde, stringSerde, DEFAULT_OUTPUT_TOPIC);
+
+        // Remove any state from previous test runs
+        IntegrationTestUtils.purgeLocalStreamsState(streamsConfiguration);
+
+        KafkaStreams streams = new KafkaStreams(builder, streamsConfiguration);
+
+        Field streamThreadsField = streams.getClass().getDeclaredField("threads");
+        streamThreadsField.setAccessible(true);
+        StreamThread[] streamThreads =  (StreamThread[]) streamThreadsField.get(streams);
+        StreamThread originalThread = streamThreads[0];
+
+        TestStreamThread testStreamThread = new TestStreamThread(builder, streamsConfig,
+                                           new DefaultKafkaClientSupplier(),
+                                           originalThread.applicationId, originalThread.clientId, originalThread.processId, new Metrics(), new SystemTime());
+
+        streamThreads[0] = testStreamThread;
+        streams.start();
+        Thread.sleep(1000);
```

Comment:
In general we want to avoid sleeping in integration test as it may introduce transient timing-dependent failures, and also making the unit test suite taking longer time. Could we use some `TestUtil` functions like `waitUtil(// addTasks is called //)` instead? Same for all other sleeping calls.